server/commitlog/message_set.go

```diff
@@ -66,6 +66,14 @@ func newMessageSetFromProto(baseOffset, basePos int64, msgs []*Message) (
 			relPos = int64(n)
 			offset = int64(i) + baseOffset
 		)
+
+		// Check expected offset for concurrency in case of Optimistic Concurrency Control
+		if concurrencyControl == true {
+			if offset != m.Offset {
```

Comment:
If the expected offset is -1, this check should be skipped right?

Reply:
WIP